### PR TITLE
IH-615: move metrics collection out of RRDD

### PR DESCRIPTION
**Pull request by last-genius**
Moves out two metrics collectors out of RRDD into separate plugins:
* CPU-related metrics
* Network-related metrics

Three more collectors are still in `xcp_rrdd.ml`, this is because memory-, HA-, and SR-cache-related metrics collection is tightly integrated with RRDD itself, exposing calls through RRDD bindings for XAPI to use. To convert these into plugins (and drop the collection-related code in `xcp_rrdd.ml`, and stop linking to Xenctrl), these calls need to be done through standard message switch mechanism instead. This is for a future PR.

Also:
* netdev metrics generation is tightly integrated with networkd, serializing these into a file, which the plugin reads from. not terribly efficient but i don't see a way to disentangle it from networkd, so it's kept this way.

Needs to be merged together with a corresponding spec repo PR, I will open it after XenRT testing completes (currently running BVT/BST) + also need to test manually

### CODE REVIEW

File: ocaml/xcp-rrdd/lib/plugin/xenctrl_lib.mli [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+val domain_snapshot :
+     Xenctrl.handle
+  -> float * (Xenctrl.domaininfo * string * int) list * string list
```
Comment (psafont) [inline ocaml/xcp-rrdd/lib/plugin/xenctrl_lib.mli:17]: please add a doc-comment explaining what the function does (provide a list of running, non-migrating vms, and paused ones)

File: ocaml/xcp-rrdd/bin/rrdd/rrdd_server.ml [hunk 2]
```diff
@@ -807,17 +807,3 @@ module Plugin = struct
   (* Read, parse, and combine metrics from all registered plugins. *)
   let read_stats () : (Rrd.ds_owner * Ds.ds) list = Local.read_stats ()
 end
-
-module HA = struct
-  let enable_and_update (statefile_latencies : Rrd.Statefile_latency.t list)
```
Comment (psafont) [inline ocaml/xcp-rrdd/bin/rrdd/rrdd_server.ml:812]: So these two functions weren't being used? lucky...

**Comment by last-genius** [comment]
Encountered crashes with the previous version of the code, determined that the cause was removing deprecated xcp-rrdd calls, so reverted these and had to also leave HA-related metrics collection as is for now, it needs to be converted to use the message switch separately. This is currently finishing up XenRT testing after some manual testing and looking good.

**Comment by psafont** [comment]
> after some manual testing

What did this involve? Opening xencenter and checking that all the metrics were still reported? (and check that they changed when adding traffic / cpu load)

**Comment by last-genius** [comment]
> What did this involve? Opening xencenter and checking that all the metrics were still reported? (and check that they changed when adding traffic / cpu load)

I initially tested with `xe host-data-source-query`, but yes, XenCenter also shows performance stats for hosts and VMs (in particular, the modified CPU and network-related stats):

{DF938305-8227-4948-96F6-59B219142C51}

**Review by psafont — APPROVED** [review]

**Review by robhoes — APPROVED** [review]
We have discussed moving the part of network_monitor_thread.ml (xcp-networkd) that ends up doing into the RRDs (the rx and tx counters) to the new RRD plugin daemon. We will leave the other stats (carrier, duplex, etc), which are read by xapi and put in its DB, in xcp-networkd.

I am approving this PR now so that we can merge it, but we should do the above relatively soon as well.

**Pull request merged**